File: services/app/server/src/textVectorizer.ts
const TOKEN_RE = /[a-z0-9]+/g;
export const DEFAULT_DIM = 384;

export function tokenizeText(text: string): string[] {
  return (text.toLowerCase().match(TOKEN_RE) ?? []).filter(Boolean);
}

function hashToken(token: string): Uint8Array {
  let hash = 2166136261;
  for (let i = 0; i < token.length; i += 1) {
    hash ^= token.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  const bytes = new Uint8Array(8);
  for (let i = 0; i < 8; i += 1) {
    bytes[i] = (hash >>> ((i % 4) * 8)) & 0xff;
  }
  return bytes;
}

export function embedText(text: string, dim = DEFAULT_DIM): number[] {
  const tokens = tokenizeText(text);
  if (!tokens.length) return Array.from({ length: dim }, () => 0);
  const vector = Array.from({ length: dim }, () => 0);
  for (const token of tokens) {
    const bytes = hashToken(token);
    const slot = (((bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3]) >>> 0) % dim;
    const sign = bytes[4] % 2 ? -1 : 1;
    const weight = 1 + Math.min(token.length, 12) / 12;
    vector[slot] += sign * weight;
  }
  const norm = Math.sqrt(vector.reduce((sum, value) => sum + value * value, 0));
  if (norm <= 1e-9) return Array.from({ length: dim }, () => 0);
  return vector.map((value) => value / norm);
}

export function vectorLiteral(values: number[]): string {
  return `[${values.map((value) => value.toFixed(8)).join(",")}]`;
}

File: src/photo_curator/text_vectorizer.py
from __future__ import annotations

import math
import re

TOKEN_RE = re.compile(r"[a-z0-9]+")
DEFAULT_DIM = 384


def tokenize_text(text: str) -> list[str]:
    return TOKEN_RE.findall(text.lower())


def embed_text(text: str, dim: int = DEFAULT_DIM) -> list[float]:
    tokens = tokenize_text(text)
    if not tokens:
        return [0.0] * dim

    vector = [0.0] * dim
    for token in tokens:
        hash_value = 2166136261
        for ch in token:
            hash_value ^= ord(ch)
            hash_value = (hash_value * 16777619) & 0xFFFFFFFF
        bytes_ = [(hash_value >> ((i % 4) * 8)) & 0xFF for i in range(8)]
        slot = (
            int(((bytes_[0] << 24) | (bytes_[1] << 16) | (bytes_[2] << 8) | bytes_[3]) & 0xFFFFFFFF)
            % dim
        )
        sign = -1.0 if bytes_[4] % 2 else 1.0
        weight = 1.0 + min(len(token), 12) / 12.0
        vector[slot] += sign * weight

    norm = math.sqrt(sum(value * value for value in vector))
    if norm <= 1e-9:
        return [0.0] * dim
    return [value / norm for value in vector]


def vector_literal(values: list[float]) -> str:
    return "[" + ",".join(f"{value:.8f}" for value in values) + "]"
